migration/run.js

```diff
@@ -1,29 +1,88 @@
+const base = `app`;
+
+glob(`${base}/**/*.ts`, {}, function(_, files) {
+  const removeOnDestroy = process.argv.includes('--removeOnDestroy');
 
-glob(`${base}/**/*.ts`, {}, function(er, files) {
   files.forEach(path => {
-    fs.readFile(path, 'utf8', function(err, text) {
-      if (hasUntilDestroy.test(text)) {
+    fs.readFile(path, 'utf8', function(_, text) {
+      if (!hasUntilDestroy.test(text)) return;
+
+      const result = transformCode(text, { removeOnDestroy });
+
+      fs.writeFile(path, result, 'utf8', function(err) {
+        if (err) return console.log(err);
         console.log(`Replaced ${path}`);
-        const result = text
-          .replace(
-            /((?:@\w*\([^]*\)[\n\s\r\t]*)(?=([\n\s\r\t]*export[\s\r\t]*class))\2)/,
-            '@UntilDestroy()\n$1'
-          )
-          .replace(
-            catchImport,
-            `import { untilDestroyed, UntilDestroy } from '@ngneat/until-destroy';`
-          );
-        fs.writeFile(path, result, 'utf8', function(err) {
-          if (err) return console.log(err);
-        });
-      }
+      });
     });
   });
 });
+
+function transformCode(str, { removeOnDestroy = false } = {}) {
+  const sourceFile = new Project().createSourceFile(`code.ts`, str);
+  replaceOldImport(sourceFile);
+
+  const classes = sourceFile.getClasses() || [];
+  classes.forEach(clazz => {
+    addUntilDestroyDecorator(clazz);
+
+    if (removeOnDestroy) {
+      const ngOnDestroyMember = clazz.getMember('ngOnDestroy');
+      const isNgOnDestroyEmpty = Boolean(
+        ngOnDestroyMember && ngOnDestroyMember.getDescendantStatements().length
+      );
+      if (isNgOnDestroyEmpty) return;
+      ngOnDestroyMember.remove();
+
+      removeOnDestroyImplements(clazz);
+      removeOnDestroyImport(sourceFile);
+    }
+  });
+
+  return sourceFile.getFullText();
+}
+
+function replaceOldImport(sourceFile) {
+  const oldImport = sourceFile.getImportDeclaration('ngx-take-until-destroy');
+  oldImport &&
+    oldImport.replaceWithText(
+      `import { UntilDestroy, untilDestroyed } from '@ngneat/until-destroy';`
+    );
+}
+
+function addUntilDestroyDecorator(classDeclaration) {
+  const decorators = [
+    { name: 'UntilDestroy', arguments: [] },
+    ...classDeclaration.getStructure().decorators
+  ];
+  classDeclaration.getDecorators().forEach(d => d.remove());
+  classDeclaration.addDecorators(decorators);
+}
+
+function removeOnDestroyImplements(classDeclaration) {
+  const onDestroyImpl = classDeclaration
+    .getImplements()
+    .find(impl => impl.getText() === 'OnDestroy');
+  classDeclaration.removeImplements(onDestroyImpl);
+}
+
+function removeOnDestroyImport(sourceFile) {
+  const importDeclaration = sourceFile.getImportDeclaration('@angular/core');
+
+  importDeclaration
+    .getImportClause()
+    .getNamedImports()
+    .find(node => node.getText() === 'OnDestroy')
+    .remove();
+
+  if (!importDeclaration.getImportClause()) {
+    importDeclaration.remove();
+  }
+}
```

Comment:
```suggestion
/**
 * @param {import('ts-morph').SourceFile} sourceFile
 */
function removeOnDestroyImport(sourceFile) {
  const importDeclaration = sourceFile.getImportDeclaration('@angular/core');

  if (importDeclaration === undefined) {
    return;
  }

  const namedImports = importDeclaration.getImportClause().getNamedImports();
  const onDestroyImportSpecifier = namedImports.find(node => node.getText() === 'OnDestroy');

  if (onDestroyImportSpecifier === undefined) {
    return;
  }

  onDestroyImportSpecifier.remove();
}
```

You're accessing the `ImportClause` at first and then you check if it's `undefined`. You'd probably just want to remove the `OnDestroy` interface import.

Reply:
I also want to remove `importDeclaration` if `ImportClause` is `undefined`

Reply:
1) It's `undefined` only for _value-less_ imports: `import 'zone.js'`
2) It would've thrown an exception `Cannot read property 'getNamedImports' of undefined` before

Reply:
@arturovt 
1. Yes, and I want to remove empty import after `OnDestroy` removing.
2. Ok, but it's rare case. I think, I'll check it in `run.ts`